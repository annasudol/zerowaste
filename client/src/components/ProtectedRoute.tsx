import * as React from "react";
import {
    RouteProps,
    Redirect,
    RouteComponentProps,
    Route
} from 'react-router-dom'
import { isLoggedIn } from '../userAuth'

interface Props extends RouteProps {
    allowed: boolean
    redirectTo: string
    component: any
    render: any
}

const ProtectedRoute: React.FunctionComponent<Props> = ({
    allowed,
    redirectTo,
    component: Component,
    render,
    children,
    ...rest
}) => (
        <Route
            {...rest}
            render={(props: RouteComponentProps) => {
                if (allowed) {
                    if (Component) {
                        return <Component {...props} />;
                    }
                    else if (render) {
                        return render(props);
                    }
                    else {
                        return children;
                    }
                }

                return <Redirect to={redirectTo} />;
            }} />
    )

export const PrivateRoute: React.FunctionComponent<RouteProps> = props => (
    <ProtectedRoute {...props} allowed={isLoggedIn()} redirectTo='/login' />
)

export const PublicRoute: React.FunctionComponent<RouteProps> = props => (
    <ProtectedRoute {...props} allowed={!isLoggedIn()} redirectTo='/home' />
)
