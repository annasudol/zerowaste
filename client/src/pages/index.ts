export * from './AddRecipeForm';
export * from './Login';
export * from './Main';
export * from './RecipeInfo';
export * from './RecipesList';
export * from './SingUp';
export * from './UserInfo';
