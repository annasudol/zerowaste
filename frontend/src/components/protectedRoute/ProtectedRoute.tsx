import React, { FC, ReactElement } from 'react';

import {
    RouteProps,
    Redirect,
    RouteComponentProps,
    Route
} from 'react-router-dom'
import { isLoggedIn } from '../../userAuth'
import { AppRoutes } from '../../routes'
    ;

interface Props extends RouteProps {
    allowed?: boolean
    redirectTo: string
    component?: any
    render?: any
};

const ProtectedRoute: FC<Props> = ({
    allowed,
    redirectTo,
    component: Component,
    render,
    children,
    ...rest
}): ReactElement => (
        <Route
            {...rest}
            render={(props: RouteComponentProps) => {
                if (allowed) {
                    if (Component) {
                        return <Component {...props} />;
                    }
                    else if (render) {
                        return render(props);
                    }
                    else {
                        return children;
                    }
                }

                return <Redirect to={redirectTo} />;
            }} />
    )

export const PrivateRoute: FC<RouteProps> = props => (
    <ProtectedRoute {...props} allowed={isLoggedIn()} redirectTo={AppRoutes.SignUp} />
)

export const PublicRoute: FC<RouteProps> = props => (
    <ProtectedRoute {...props} allowed redirectTo={AppRoutes.Home} />
)
