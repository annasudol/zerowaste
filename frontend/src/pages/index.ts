export * from './addRecipeForm/AddRecipeForm';
export * from './editRecipeForm/EditRecipeForm';
export * from './login/Login';
export * from './home/Home';
export * from './recipeInfo/RecipeInfo';
export * from './recipeList/RecipesList';
export * from './signUp/SingUp';
export * from './userInfo/UserInfo';