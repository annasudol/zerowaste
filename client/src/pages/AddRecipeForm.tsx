import * as React from "react";
import { ListUI, ErrorMessage, ListAddForm, AutocompleteIngredients } from '../components'
import { useStepsState, useDetailedIngredientState, useRecipeFormState, initialState } from "../hooks";
import { useHistory } from 'react-router';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/react-hooks';

const ADD_RECIPE = gql`
  mutation AddRecipe($title: String! $image: String $readyInMinutes: Int! $ingredients: [String!]! $detailedIngredients: [String!]! $steps: [String!] $sourceUrl: String $author: String $authorId: String) {
    createRecipe(title: $title, image: $image, readyInMinutes: $readyInMinutes, ingredients: $ingredients, detailedIngredients: $detailedIngredients, steps: $steps, sourceUrl: $sourceUrl, author: $author, authorId: $authorId) {
        id
        title
    }
  }

`;

export const AddRecipeForm: React.FunctionComponent = (): React.ReactElement => {
    const { detailedIngredients, addDetailedIngredient, deleteDetailedIngredient } = useDetailedIngredientState();
    const { steps, addStep, deleteStep } = useStepsState();
    const { title, ingredients, readyInMinutes, sourceUrl, dispatch } = useRecipeFormState();
    const [emptyInput, setEmptyInput] = React.useState<boolean>(false);
    const [createRecipe, { data, error }] = useMutation(ADD_RECIPE);
    const history = useHistory();

    const handleSubmit = (event): any => {

        event.preventDefault();

        if (initialState.title || initialState.readyInMinutes || ingredients.length === 0 || detailedIngredients.length === 0 || steps.length === 0) {
            setEmptyInput(true)
        } else {
            setEmptyInput(false)
            createRecipe({ variables: { title, image: 'https://spoonacular.com/recipeImages/543832-556x370.jpg', readyInMinutes, ingredients, detailedIngredients, steps, sourceUrl, author: 'kevin', authorId: "2" } });
            history.push({ pathname: `/recipe/${data.createRecipe.id}` });
        }

    }


    if (error) {
        return (<ErrorMessage message='Error with adding recipe, please try later' />)
    }
    return (
        <div className="content p-10">
            <div className="flex justify-center items-center">
                <div className="flex flex-col w-1/2 recipe-form">
                    <h1 className="form-header font-bebas uppercase text-darkGray text-center pb-0 m-0">Add Recipe</h1>
                    <form onSubmit={(e): any => handleSubmit(e)}>
                        {emptyInput && title === '' && <ErrorMessage validationMessage='Add title' />}
                        <input name="title" type="text" placeholder="Title" className="mt-2 mb-2" value={title} onChange={(e: React.ChangeEvent<HTMLInputElement>): void => dispatch({ TYPE: 'ADD_TITLE', setTitle: e.target.value })} />
                        {emptyInput && readyInMinutes === 0 && <ErrorMessage validationMessage='Add preparation time' />}
                        <input name="readyInMinutes" type="number" placeholder="Ready In Minutes" className="mt-2 mb-2" value={readyInMinutes}
                            onChange={(e: React.ChangeEvent<HTMLInputElement>): void => {
                                const val = parseInt(e.target.value, 10);
                                if (val > 0) dispatch({ TYPE: 'ADD_READY_IN_MINUTES', setMinutes: val })
                            }} />
                        {emptyInput && ingredients.length === 0 && <ErrorMessage validationMessage='Add ingredients to make the recipe searchable' />}
                        <AutocompleteIngredients dispatch={dispatch} ingredients={ingredients} />

                        {emptyInput && detailedIngredients.length === 0 && <ErrorMessage validationMessage='Add detailed ingredients eg: 1 tbsp olive oil' />}

                        <ListAddForm saveItem={(text: string) => {
                            const trimmedText = text.trim();
                            if (trimmedText.length > 0) {
                                addDetailedIngredient(trimmedText);
                            }
                        }}
                            placeholder="Add detailed ingredients"
                        />
                        <ListUI list={detailedIngredients} deleteItem={(index: number): any => deleteDetailedIngredient(index)} />
                        {emptyInput && steps.length === 0 && <ErrorMessage validationMessage='Add recipe steps' />}
                        <ListAddForm saveItem={(text: string) => {
                            const trimmedText = text.trim();
                            if (trimmedText.length > 0) {
                                addStep(trimmedText);
                            }
                        }}
                            placeholder="Add steps eg.: Preheat oven to 350ยบ. Rinse rice under cold water until the water runs clear. Bring a large pot of water to a boi"
                            textarea={true}
                        />
                        <ListUI list={steps} deleteItem={(index: number): any => deleteStep(index)} />
                        <input name="sourceUrl" type="text" placeholder="Source Url" value={sourceUrl} onChange={(e: React.ChangeEvent<HTMLInputElement>): void => dispatch({ TYPE: 'ADD_SOURCE_URL', setUrl: e.target.value })} />
                        <button type="submit" className="bg-green hover:bg-green text-white font-bold py-2 px-4 rounded outline-none m-1 border-0">Send Recipe</button>
                    </form>
                </div>
            </div>
        </div>
    )
};