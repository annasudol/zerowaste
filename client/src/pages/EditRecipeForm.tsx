import * as React from "react";
import { useLocation, useHistory, useParams } from 'react-router';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/react-hooks';
import { Redirect } from "react-router";
import BackspaceIcon from '@material-ui/icons/Backspace';
import { RecipeForm } from "../components";
import { RecipeStateProps } from "../utils/types";
import { Button } from 'antd';
import { AppRoutes } from "../../routes";
import { Store } from 'antd/lib/form/interface';

const UPDATE_RECIPE = gql`
  mutation UpdateRecipe($id: ID! $title: String! $servings: Int! $image: String!, $readyInMinutes: Int! $ingredients: [String!]! $detailedIngredients: [String!]! $instructions: String! $sourceUrl: String) {
    updateRecipe(id: $id title: $title, servings: $servings image: $image, readyInMinutes: $readyInMinutes, ingredients: $ingredients, detailedIngredients: $detailedIngredients, instructions: $instructions, sourceUrl: $sourceUrl) {
        id
    }
  }
`;

export interface LocationTypes {
    hash?: string
    pathname: string
    search?: string
    state: RecipeStateProps
    key?: string
}


export const EditRecipeForm: React.FC = (): React.ReactElement => {
    const location: LocationTypes = useLocation();
    const initialState: RecipeStateProps = location.state;
    const { recipeID } = useParams();
    const [updateRecipe, { data }] = useMutation(UPDATE_RECIPE);
    const history = useHistory();

    const handleSubmit = (inputValues: Store): any => {
        const { title, servings, image, readyInMinutes, ingredients, detailedIngredients, instructions, sourceUrl } = inputValues;
        updateRecipe({ variables: { id: recipeID, title, servings, image, readyInMinutes, ingredients, detailedIngredients, instructions, sourceUrl } });
    }

    if (data) {
        return <Redirect to={{ pathname: `/recipe/${data.updateRecipe.id}`, state: { backPath: '/user', callRefetch: true } }} />
    }
    return (
        <div className="content">
            <Button type="link" size="large" className="mt-4" onClick={(): any => history.push({ pathname: AppRoutes.User })}> <BackspaceIcon color="error" /></Button>
            <div className="flex justify-center items-center">
                <div className="w-1/2 recipe-form pb-4 pt-4">
                    <h1 className="form-header font-bebas uppercase text-darkGray text-center pb-0 m-0">Edit Recipe</h1>
                    <RecipeForm handleSubmit={(inputValues: Store): void => handleSubmit(inputValues)} defaultValues={initialState} />
                </div>
            </div>
        </div>
    )
};