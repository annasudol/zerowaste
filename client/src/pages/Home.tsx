import * as React from 'react';
import { Image, SearchRecipesForm } from '../components';
const fridge = require('./../assets/images/svg/technology.svg');

export const Home: React.FC = () => (
    <div className='content overflow-hidden'>
        <div className='flex flex-col justify-center items-center md:flex-row-reverse items-center h-screen'>
            <div className='mb-8 text-center md:w-1/3 lg:w-1/4 md:text-left min-h-32'>
                <h1 className='font-bebas uppercase text-darkGray mb-3'>your spare food</h1>
                <SearchRecipesForm />
            </div>
            <Image src={fridge.default} alt='fridge' size='medium' />
        </div>
    </div>)