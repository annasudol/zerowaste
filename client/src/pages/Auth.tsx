import * as React from "react";
import { useForm } from "react-hook-form";
import { Button } from "../components";
import { useApolloClient, useMutation } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import { useHistory } from 'react-router';
import ApolloClient from 'apollo-client';
import { AppRoutes } from "../../routes";
import { Main } from "../pages";
import { Route, Redirect } from "react-router-dom";
export const LOGIN_USER = gql`
  mutation Login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
        token
    }
  }
`;

// export const SIGNUP_USER = gql`
//   mutation signup($input: {name: String!, email: String!, password: String!}) {
//     signup(input: $input) {
//         name
//         email
//         password
//     }
//   }
// `;
type Inputs = {
    example: string,
    exampleRequired: string,
};


// export interface LoginProps {
//     login: string;
//   }




export const Auth: React.FunctionComponent = (): React.ReactElement => {
    const history = useHistory();

    const client: ApolloClient<any> = useApolloClient();
    const { register, handleSubmit, watch, errors } = useForm<Inputs>();
    const [loginPage, setLoginPage] = React.useState(false);

    const [login, { loading, error }] = useMutation(
        LOGIN_USER,
        {
            // tslint:disable-next-line: no-shadowed-variable
            onCompleted({ login }) {
                localStorage.setItem('token', login.token);
                client.writeData({ data: { isLoggedIn: true } });
            }
        }
    );

    const toggleSetLoginPage = (): void => {
        setLoginPage(!loginPage);
    }

    const submit = (data: any) => {
        if (loginPage) {
            console.log(data)
            login({ variables: { email: data.email, password: data.password } });
            return history.push({ pathname: `${AppRoutes.Home}` });
            // return <Redirect to={{ pathname: `${AppRoutes.Home}` }} />
        }

    }

    return (
        <div className="content overflow-hidden flex justify-center items-center">
            <div className="form  flex flex-col">
                <h1 className="form-header font-bebas uppercase text-darkGray text-center pb-0 m-0">{loginPage ? 'Log in' : 'Create an account'}</h1>
                <p className="font-roboto text-center mb-4">{!loginPage ? 'Already have an account?' : "Don't have an account?"}</p>
                <Button onClick={toggleSetLoginPage} color="coral" className="mb-4">{!loginPage ? 'Switch to Log in' : 'Switch to Sign in'}</Button>
                <form onSubmit={handleSubmit(submit}>
                    {!loginPage && <input name="username" type="text" placeholder="User Name" ref={register({ required: true, maxLength: 30 })} />}
                    <input
                        name="email"
                        placeholder="Email"
                        type="email"
                        ref={register({
                            required: true,
                            pattern: {
                                value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
                                message: "invalid email address"
                            }
                        })}
                    />
                    <input name="password" type="password" placeholder="Set a password" ref={register({ required: true })} />
                    <input type="submit" title="Submit" className="bg-green hover:bg-green text-white font-bold py-2 px-4 rounded outline-none m-1 border-0" />
                </form>
            </div>
        </div>)

}
