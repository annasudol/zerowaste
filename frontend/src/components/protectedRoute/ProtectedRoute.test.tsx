import { shallow } from 'enzyme';
import React from 'react';
import { PrivateRoute, PublicRoute } from './ProtectedRoute';
import toJson from 'enzyme-to-json';
import { Login } from '../../pages';


describe('component', (): void => {
  describe('LoginSignUpForm', (): void => {
    describe('match snapshot', (): void=> {
        it('PrivateRoute', (): void => {
            const wrapper = shallow(<PrivateRoute component={Login} path="newPath" />);
            expect(toJson(wrapper)).toMatchSnapshot();
          });
          it('PublicRoute', (): void => {
            const wrapper = shallow(<PublicRoute component={Login} path="newPath" />);
            expect(toJson(wrapper)).toMatchSnapshot();
          });
    });
  });
});