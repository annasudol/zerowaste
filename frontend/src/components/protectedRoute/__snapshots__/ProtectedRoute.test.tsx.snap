// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component LoginSignUpForm match snapshot PrivateRoute 1`] = `
<ProtectedRoute
  allowed={false}
  component={[Function]}
  path="newPath"
  redirectTo="/signUp"
/>
`;

exports[`component LoginSignUpForm match snapshot PublicRoute 1`] = `
<ProtectedRoute
  allowed={true}
  component={[Function]}
  path="newPath"
  redirectTo="/"
/>
`;
